apacheconfig/parser.py

```diff
@@ -74,50 +74,93 @@ def parse(self, text):
         self.reset()
         return self.engine.parse(text)
 
-    # Parsing rules
+    # PARSING RULES
+    # =============
+
+    def p_requirednewline(self, p):
+        """requirednewline : NEWLINE
+        """
+        p[0] = p[1:][0]
+
+    def p_whitespace(self, p):
+        """whitespace : requirednewline
+                      | WHITESPACE
+        """
+        p[0] = p[1:][0]
 
     def p_statement(self, p):
         """statement : OPTION_AND_VALUE
+                     | OPTION_AND_VALUE_NOSTRIP
         """
-        p[0] = ['statement', p[1][0], p[1][1]]
+        p[0] = ['statement', p[1][0], p[1][2]]
 
         if self.options.get('lowercasenames'):
             p[0][1] = p[0][1].lower()
 
-        if (not self.options.get('nostripvalues') and
-                not hasattr(p[0][2], 'is_single_quoted') and
-                not hasattr(p[0][2], 'is_double_quoted')):
-            p[0][2] = p[0][2].rstrip()
+    # Note: item vs comment
+    # ---------------------
+    # `item` and `comment` are differentiated since there can be in-line
+    # comments. Comments do not necessarily need a newline to separate it
+    # from a previous item, but an item needs a newline to
+    # separate it from a previous item or comment.
 
     def p_item(self, p):
         """item : statement
-                | comment
                 | include
                 | includeoptional
                 | block
         """
         p[0] = p[1:][0]
 
+    def p_startitem(self, p):
+        """startitem : whitespace item
+                     | whitespace comment
+                     | item
+                     | comment
+        """
+        if len(p) == 3:
+            p[0] = p[1:][1]
+        else:
+            p[0] = p[1:][0]
+
+    def p_miditem(self, p):
+        """miditem : requirednewline item
+                   | whitespace comment
+                   | comment
+        """
+        if len(p) == 3:
+            p[0] = p[1:][1]
+        else:
+            p[0] = p[1:][0]
+
     def p_contents(self, p):
-        """contents : contents item
-                    | item
+        """contents : contents miditem
```

Comment:
nit: It feels like this BNF can be slightly reduced.

Reply:
Any suggestions? Also happy to explain some of the choices here in the comments, like splitting startitem/miditem!

Reply:
That's a nitpick anyway, let's push this into a TODO and get back to this later?

Reply:
Sounds good! Tests are passing now-- let's merge this in?